File: zy70201/backend/src/controllers/reportController.ts
import { Request, Response } from 'express';
import { reportService, CreateReportDto } from '../services/reportService';
import { handleError } from '../utils/errorHandler';

export const reportController = {
  async getAllReports(req: Request, res: Response) {
    try {
      const reports = await reportService.getAllReports();
      res.json({ success: true, data: reports });
    } catch (error) {
      handleError(res, error);
    }
  },

  async getReportById(req: Request, res: Response) {
    try {
      const report = await reportService.getReportById(req.params.id);
      res.json({ success: true, data: report });
    } catch (error) {
      handleError(res, error);
    }
  },

  async createReport(req: Request, res: Response) {
    try {
      const data: CreateReportDto = req.body;
      const report = await reportService.createReport(data);
      res.status(201).json({ success: true, data: report });
    } catch (error) {
      handleError(res, error);
    }
  },

  async approveReport(req: Request, res: Response) {
    try {
      const { approver } = req.body;
      const report = await reportService.approveReport(req.params.id, approver);
      res.json({ success: true, data: report });
    } catch (error) {
      handleError(res, error);
    }
  },

  async exportReports(req: Request, res: Response) {
    try {
      const format = (req.query.format as 'json' | 'excel') || 'json';
      const startDate = req.query.startDate as string;
      const endDate = req.query.endDate as string;
      
      const result = await reportService.exportReports(format, startDate, endDate);
      
      if (format === 'excel') {
        res.setHeader('Content-Type', 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet');
        res.setHeader('Content-Disposition', `attachment; filename="reports-${Date.now()}.xlsx"`);
        res.send(result);
      } else {
        res.json({ success: true, data: result });
      }
    } catch (error) {
      handleError(res, error);
    }
  }
};

File: zy71023/src/main/java/com/hazardous/waste/controller/ReportController.java
package com.hazardous.waste.controller;

import com.hazardous.waste.dto.ApiResponse;
import com.hazardous.waste.entity.DisposalReport;
import com.hazardous.waste.entity.WasteRecord;
import com.hazardous.waste.enums.WasteStatus;
import com.hazardous.waste.service.ReportService;
import com.hazardous.waste.service.WasteRecordService;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.io.IOException;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/reports")
public class ReportController {

    private ReportService reportService;
    private WasteRecordService wasteRecordService;

    @PostMapping("/disposal/generate")
    public ApiResponse<DisposalReport> generateDisposalReport(
            @RequestParam String transferFormNo,
            @RequestParam(defaultValue = "SYSTEM") String operator) {
        return ApiResponse.success(reportService.generateDisposalReport(transferFormNo, operator));
    }

    @PostMapping("/disposal/{reportNo}/approve")
    public ApiResponse<DisposalReport> approveReport(
            @PathVariable String reportNo,
            @RequestBody Map<String, Object> request) {
        String approver = (String) request.getOrDefault("approver", "SYSTEM");
        Boolean approved = (Boolean) request.getOrDefault("approved", true);
        String remark = (String) request.get("remark");
        return ApiResponse.success(reportService.approveReport(reportNo, approver, approved, remark));
    }

    @GetMapping
    public ApiResponse<List<DisposalReport>> getAllReports() {
        return ApiResponse.success(reportService.getAllReports());
    }

    @GetMapping("/{reportNo}")
    public ApiResponse<DisposalReport> getReportByNo(@PathVariable String reportNo) {
        return ApiResponse.success(reportService.getReportByNo(reportNo));
    }

    @GetMapping("/export/waste-records")
    public ResponseEntity<byte[]> exportWasteRecords(
            @RequestParam(required = false) WasteStatus status) throws IOException {
        List<WasteRecord> records = status != null
                ? wasteRecordService.getRecordsByStatus(status)
                : wasteRecordService.getAllRecords();

        byte[] excelData = reportService.exportWasteRecordsToExcel(records);

        String filename = "危废记录_" + LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyyMMdd")) + ".xlsx";

        return ResponseEntity.ok()
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=\"" + filename + "\"")
                .contentType(MediaType.parseMediaType("application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"))
                .body(excelData);
    }

    @GetMapping("/export/disposal/{reportNo}")
    public ResponseEntity<byte[]> exportDisposalReport(@PathVariable String reportNo) throws IOException {
        DisposalReport report = reportService.getReportByNo(reportNo);
        byte[] excelData = reportService.exportDisposalReportToExcel(report);

        String filename = "处置报告_" + reportNo + ".xlsx";

        return ResponseEntity.ok()
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=\"" + filename + "\"")
                .contentType(MediaType.parseMediaType("application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"))
                .body(excelData);
    }
}
